### DISCUSSION TITLE
Assigning loose "unknown" files to a project and disabling unknown category?

### DISCUSSION THREAD

**Discussion by gingerbeardman** [muety/wakapi#617] [Q&A]
At times I seem to open up single files from within projects, and these are sorted as "unknown"

But a regex of the file path is enough to assign the project.

It's odd to see these loose files as "unknown" because there is a matching project. 

eg.
`/Users/matt/Projects/Playdate/sharkturtle/Source/main.lua`
should be part of
`sharkturtle`

Is there any way I can manually assign these? 

Preferably with the web interface, or if command line is required I'd need step-by-step help. 🆘 

There are a dozen or two of them I would really like to remap. But there are almost 2000 of them in total. 😱 

After remapping the most important ones, I would want to disable the "unknown" category. Is this possible?

Thanks for a useful project.

**Comment by gingerbeardman** [comment]
On wakatime I can do this with custom rules, and they are applied to past coding activity. 

https://wakatime.com/faq#custom-rules

`(if) file (contains) sharkturtle (then change) project (to) sharkturtle`

This might be a workaround for the tidy up, after which I can reimport all my data. I have to wait for 24 hours before testing this, as I already did an import today.

I notice custom rules in wakapi do not seem to offer the same options here.

**Comment by gingerbeardman** [comment]
Alternatively, another way would be the ability to ingest `~/.wakatime.cfg` and apply projectmap settings retrospectively.

**Comment by muety** [comment]
Hi @gingerbeardman, if I understood it right, the _project_ is labeled as _unknown_, but not the _language_, correct? But it only happens for certain file within the same project folder structure? That's weird, don't really have a good explanation for that 🤔.

Generally, a good practice is to place a `.wakatime-project` file (see [here](https://github.com/wakatime/wakatime-cli/blob/develop/USAGE.md#project-detection)) inside your project.

In case Wakapi's current rules functionality is not sufficient, there is #607 to add support for regexes. I'll work on implementing that soon.

If it's only about a one-time change to your already existing data, you could either do another import as you suggested or, if you're using wakapi.dev, send me an e-mail and tell me what remappings to perform via some manual, one-off database queries.

Hope that helps.

**Reply by gingerbeardman** [reply]
Going forward I'll use the config. 

I guess I've opened 2000 files that don't belong to a project, or aren't picked up as part of the project. 

The example above is odd because the I do have a project being tracked for it so I'm not sure why this one file isn't incuded. Maybe the combination of editor and opening as a single file rather than a folder?

I'm happy to provide data if you want to look at it. 

I'll post screenshots after my next import, tomorrow.

**Reply by muety** [reply]
> Maybe the combination of editor and opening as a single file rather than a folder?

Yes, I think if you open a single file in VSCode, it will be tracked as unknown project.

> I'm happy to provide data if you want to look at it.

Please do!

**Reply by gingerbeardman** [reply]
Here's a JSON export from my Wakatime before I made any changes using custom rules.
https://www.dropbox.com/scl/fi/55gtntxt6l8yf6n1909yp/wakatime-3fe3a916ed7045b794a4a9e53d9367a0.json.7z?rlkey=jbcnlz9wqax50qvhxita0lpme&dl=0

I also did an import at gingerbeardman on wakapi.dev but this is with data that I have been modifiying using custom rules to try to assign many of the loose files.

**Reply by gingerbeardman** [reply]
screenshots of the data currently imported into wakapi.dev
(screenshot taken from my local self-hosted mirror)

all my custom rules only matched 10% of unknown files.

![Screen shot 2024-03-10 at 15 31 49](https://github.com/muety/wakapi/assets/49612/13c4a025-7a91-430b-b9b7-dc1c8bc6f461)

![Screen shot 2024-03-10 at 15 32 40](https://github.com/muety/wakapi/assets/49612/99e9feb2-87fe-4026-b899-69a6baa14088)

**Reply by muety** [reply]
Will have a deeper look at latest on the weekend!

> all my custom rules only matched 10% of unknown files.

What do you mean by that?

**Reply by gingerbeardman** [reply]
i tried to assign these loose/unknown files by settings lots of custom rules at wakatime. 

even with dozens of custom rules I was only able to match and clean up a small number of such files. so, 

all this to say, it's not really a workable solution.

![image](https://github.com/muety/wakapi/assets/49612/62e53e8e-686c-4358-82fd-22e07fd3f60b)

**Reply by muety** [reply]
Do you think #607 would help?

**Reply by gingerbeardman** [reply]
That would allow the same sort of tweaking as wakatime custom rules. It would help but it's not a solution for me as there are simply too many "unknown" files so I would be stuck setting regex for a long time! 

And even then there would be a remainder of unknown files that are not part of any project, truly unknown files that i'm not interested in.

**Comment by muety** [comment]
> [...] I would want to disable the "unknown" category. Is this possible?

That's not supported currently.

**Reply by gingerbeardman** [reply]
OK, have opened #619

**Comment by muety** [comment]
Will close this discussion, as there's nothing left to do (except for #619) at this point, I think?

**Reply by gingerbeardman** [reply]
OK. I would say #607 and #619 will provide adequate workarounds. Time will tell if they're enough!
